### ISSUE TITLE
tests cannot use NgHttp2.dll

### ISSUE THREAD

**Issue by gvanem** [nghttp2/nghttp2#677]
It is IMHO a good test if a test-program can use the dynamic library of a package to be tested. But this is impossible. At least on Windows. Due to a lot of functions in e.g. `nghttp2_session.c` not marked as `dllexport`. And also `nghttp2_enable_strict_preface` is assumed to come from a static lib. It would be nice to patch it like so:

```diff
--- a/main.c 2016-09-05 10:47:30
+++ b/main.c 2016-09-05 12:05:50
@@ -41,7 +41,11 @@
 #include "nghttp2_helper_test.h"
 #include "nghttp2_buf_test.h"

-extern int nghttp2_enable_strict_preface;
+extern
+#if !defined(NGHTTP2_STATICLIB) && defined(_WIN32)
+  __declspec(dllimport)
+#endif
+  int nghttp2_enable_strict_preface;

 static int init_suite1(void) { return 0; }
```

Why is a static lib required?

**Comment by tatsuhiro-t** [comment]
We only marked public function as "export".  The remaining functions are private, and therefore not exported from DLL.  To test those functions, we use static library.
Basically, static lib and dynamic one are the same, and we don't lose anything when we use static library for testing.  Am I missing something?

**Comment by gvanem** [comment]
> We only marked public function as "export". 

Is there a different naming for *public* and *private* symbols? AFAICS, no. Ref. libcurl with `curl_x` (public) and `Curl_x` (private).

> and we don't lose anything when we use static library for testing.

Yes, we do. We need to relink test-programs when a static-lib changes. Maybe not a big problem, but not really needed. But it's pretty easy to use `NgHttp2.dll` for `test.exe` and `fail.exe` anyway. By adding a:
```
#include <nghttp2/nghttp2.h>
```
in some places and use `extern NGHTTP2_EXTERN` as needed.

**Comment by tatsuhiro-t** [comment]
> Is there a different naming for public and private symbols? AFAICS, no. Ref. libcurl with curl_x (public) and Curl_x (private).

You are correct.  We use same naming for public and private.  Actually, we had not set visibility before.  We set visibility when we dealt with Windows DLL export stuff.  We just marked export for public symbols.  Perhaps, that is the problem.  It is a pain to mark everything export at the moment, is there any way to make all functions export on windows without messing all functions with dllexport?

> Yes, we do. We need to relink test-programs when a static-lib changes. Maybe not a big problem, but

Usually, when static-lib changes, test program may also change.  Then it would be safer to just build everything form scratch.

**Comment by gvanem** [comment]
> is there any way to make all functions export on windows without messing all functions with dllexport?

I often use things like:
```
nm  --defined-only *.obj | grep ' T .*_nghttp2_'  | sed 's/^.* _/  /' > nghttp2.def
```
But I'm not sure one can assume MSVC users have Msys/Cygwin installed.
